scipy/sparse/_coo.py

```diff
@@ -722,16 +810,463 @@ def _matmul_multivector(self, other):
+        if self.ndim == 1:
+            self = self.reshape((1, self.shape[0])) # prepend 1 to shape
+            self_is_1d = True
+
+        if other.ndim == 1:
+            other = other.reshape((other.shape[0], 1)) # append 1 to shape
+            other_is_1d = True
+
+        if self.shape[-1] != other.shape[-2]:
+                raise ValueError(f"shapes {self.shape} and {other.shape}"
+                                 " are not aligned for n-D dot")
+
+        new_shape_self = self.shape[:-1] + (1,) * (len(other.shape) - 1) \
+            + self.shape[-1:]
+        new_shape_other = (1,) * (len(self.shape) - 1) + other.shape
+
+        result_shape = self.shape[:-1] + other.shape[:-2] + other.shape[-1:]
+        result = self.reshape(new_shape_self) @ other.reshape(new_shape_other)
+        prod_arr = result.reshape(result_shape)
+
+        # reshape back if a or b were originally 1-D
+        if self_is_1d:
+            prod_arr = prod_arr.reshape(result_shape[1:])
+        if other_is_1d:
+            prod_arr = prod_arr.reshape(result_shape[:-1])
+
+        return prod_arr
+
+    def tensordot(self, other, axes=2):
+        if not isdense(other) and not issparse(other):
+            # If it's a list or whatever, treat it like an array
+            other_array = np.asanyarray(other)
+
+            if other_array.ndim == 0 and other_array.dtype == np.object_:
+                # Not interpretable as an array; return NotImplemented so that
+                # other's __rmatmul__ can kick in if that's implemented.
+                return NotImplemented
+
+            try:
+                other.shape
+            except AttributeError:
+                other = other_array
+
+        axes_self, axes_other = _process_axes(self.ndim, other.ndim, axes)
+
+        # Adjust negative axes for self
+        axes_self = [axis + self.ndim if axis < 0 else axis for axis in axes_self]
+        # Adjust negative axes for other
+        axes_other = [axis + other.ndim if axis < 0 else axis for axis in axes_other]
+
+        # Check for shape compatibility along specified axes
+        if any(self.shape[ax] != other.shape[bx]
+               for ax, bx in zip(axes_self, axes_other)):
+            raise ValueError("sizes of the corresponding axes must match")
+        
+        if isdense(other):
+            return self._dense_tensordot(other, axes_self, axes_other)
+        else:
+            return self._sparse_tensordot(other, axes_self, axes_other)
+
+
+    def _sparse_tensordot(self, other, axes_self, axes_other):
+        ndim_self = len(self.shape)
+        ndim_other = len(other.shape)
+
+        # Prepare the tensors for tensordot operation       
+        # Ravel non-reduced axes coordinates
+        self_non_red_coords = _ravel_non_reduced_axes(self.coords, self.shape,
+                                                      axes_self)
+        self_reduced_coords = np.ravel_multi_index(np.array(self.coords)[axes_self, :],
```

Comment:
Here is another place where we can avoid copying `coords` at line 1000
`np.array(self.coords)[axes_self, :]` can be done with `[self.coords[ax] for ax in axes_self]`
(or use a tuple instead of a list)
```suggestion
        self_reduced_coords = np.ravel_multi_index([self.coords[ax] for ax in axes_self],
```